### PR TITLE
feat: subscription trie

### PR DESCRIPTION
**Pull request by literally-user**

### CODE REVIEW

File: src/zmqtt/_internal/protocol.py [hunk 1]
```diff
@@ -102,38 +71,23 @@ def __init__(
...
         self._state = state
         self._keepalive = keepalive
         self._ping_timeout = ping_timeout
-        self._connect_timeout = connect_timeout
         self._version: Final = version
         self._buf = PacketBuffer(version=version)
         self._ping_waiters: list[asyncio.Future[None]] = []
         self._disconnecting = False
         self.started_event = asyncio.Event()
+        self._subscription_index: SubscriptionIndex | None = None
```
Comment (borisalekseev) [inline src/zmqtt/_internal/protocol.py]: I don't think this attribute is necessary. As a result, access to the index is done differently in various places.

File: src/zmqtt/_internal/protocol.py [hunk 2]
```diff
@@ -253,32 +207,41 @@ async def subscribe(
...
+                actual_filter = _shared_filter_to_actual(f)
                 new_entries[f] = SubscriptionEntry(
                     queue=asyncio.Queue(),
                     auto_ack=auto_ack,
-                    actual_filter=_shared_filter_to_actual(f),
+                    actual_filter=actual_filter,
                 )
+                self._state.subscription_index.add(actual_filter, new_entries[f])
+
         self._state.subscriptions.update(new_entries)
```
Comment (borisalekseev) [inline src/zmqtt/_internal/protocol.py]: sub is registered twice here

File: tests/test_topic_matching.py [hunk 3]
```diff
@@ -45,3 +49,28 @@ def test_filter_specificity_hash() -> None:
 
 def test_filter_specificity_bare_hash() -> None:
     assert _filter_specificity("#") == (2,)
+
+
+def test_subscription_index_returns_best_match() -> None:
```
Comment (borisalekseev) [inline tests/test_topic_matching.py:75]: Please, cover more corner cases

File: src/zmqtt/_internal/state.py [hunk 4]
```diff
@@ -70,6 +71,44 @@ class SubscriptionEntry:
...
+
+    def get(self, key: str, default: SubscriptionEntry | None = None) -> SubscriptionEntry | None:
+        return self._items.get(key, default)
+
+    def add(self, key: str, value: SubscriptionEntry) -> None:
+        existing = self._items.get(key)
+        if existing is not None:
+            self._index.remove(existing.actual_filter, existing)
+        self._items[key] = value
+        self._index.add(value.actual_filter, value)
```
Comment (borisalekseev) [inline src/zmqtt/_internal/state.py]: Is there a need for both a registry and an index? They seem to serve the same function.

File: src/zmqtt/_internal/subscription_index.py [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+from itertools import chain
+from typing import TYPE_CHECKING
+
+from zmqtt._internal.topic_matching import _segment_rank, _topic_matches
+
+if TYPE_CHECKING:
+    from zmqtt._internal.state import SubscriptionEntry
+
+
+@dataclass(slots=True)
```
Comment (borisalekseev) [inline src/zmqtt/_internal/subscription_index.py:17]: frozen=True?

File: src/zmqtt/_internal/subscription_index.py [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+class _Node:
+    children: dict[str, _Node] = field(default_factory=dict)
+    wildcard_children: dict[str, _Node] = field(default_factory=dict)
+    entries: list[tuple[str, SubscriptionEntry]] = field(default_factory=list)
+
+
+class SubscriptionIndex:
```
Comment (borisalekseev) [inline src/zmqtt/_internal/subscription_index.py:31]: Ensure that each method in this class has a single responsibility, and minimize the use of if-statements that depend on the arguments or keyword arguments passed to the methods.

File: src/zmqtt/_internal/subscription_index.py [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+        for _, entry in self._iter_entries():
+            if self._entry_matches(entry, filter_):
+                return entry
+        return default
+
+    def remove(self, filter_: str, entry: SubscriptionEntry | None = None) -> None:
+        if entry is not None:
+            self._remove_entry(filter_.split("/"), filter_, entry, self._root)
+            return
+        self._remove_matching(filter_, self._root)
```
Comment (borisalekseev) [inline src/zmqtt/_internal/subscription_index.py]: Why should we remove matching? 3.10.4 UNSUBSCRIBE Actions on both 5.0 and 3.1.1 specifications says: 

> The Topic Filters (whether they contain wildcards or not) supplied in an UNSUBSCRIBE packet MUST be compared character-by-character with the current set of Topic Filters held by the Server for the Client. If any filter matches exactly then its owning Subscription MUST be deleted [MQTT-3.10.4-1], otherwise no additional processing occurs.

Write real-broker test-case for this behaviour:
1) client subscribes on `/test/+` then on `test/concrete_id`
2) UNSUB on `/test/+`
3) publish to `/test/concrete_id`
4) assert that message received in concrete subscription

**Review by Sehat1137 — APPROVED** [review]
